File: python/taflow/metrics/longest_winning_streak.py
"""Longest consecutive winning-observation metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import LongestWinningStreak as _Native
from ._input import as_metric_series


class LongestWinningStreak:
    """Compute the longest consecutive run of strictly positive observations.

    A win is strictly greater than zero. Both exact zero and a negative value
    terminate the current streak. The independent oracle is QuantStats 0.0.81
    ``consecutive_wins`` with no aggregation and ``prepare_returns=False``.
    TAFlow returns ``None`` during empty/all-missing warm-up and ``0.0`` for a
    non-empty sample with no win. With ``nan_policy="omit"``, missing values are
    removed and therefore do not break a run; ``"raise"`` rejects them.

    ``from_returns`` consumes decimal simple period returns. ``from_pnl``
    consumes raw, non-cumulative period P&L without initial-capital conversion,
    and ``from_trades`` consumes realized P&L for closed trades. No domain is
    annualized. Infinities are rejected, while return-domain observations below
    -100% are invalid. ``append``, ``extend``, and ``reset`` are fluent. Native
    bulk work releases the GIL, and Rust owns bounded allocation-free O(1)
    streaming state.
    """

    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(self, returns: Any, *, column: str | None = None) -> "LongestWinningStreak":
        """Append chronological returns observations and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_pnl(self, pnl: Any, *, column: str | None = None) -> "LongestWinningStreak":
        """Append chronological pnl observations and return this metric."""
        self._state.from_pnl(as_metric_series(pnl, column=column))
        return self

    def from_trades(self, trades: Any, *, column: str | None = None) -> "LongestWinningStreak":
        """Append chronological trades observations and return this metric."""
        self._state.from_trades(as_metric_series(trades, column=column))
        return self

    def append(self, value: float) -> "LongestWinningStreak":
        """Append one value in the selected domain and return this metric."""
        self._state.append(float(value))
        return self

    def extend(
        self, values: Any, *, column: str | None = None
    ) -> "LongestWinningStreak":
        """Append chronological observations and return this metric."""
        self._state.extend(as_metric_series(values, column=column))
        return self

    @property
    def value(self) -> float | None:
        """Return the longest winning run, or ``None`` while empty."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current scalar without replaying processed input."""
        return self._state.compute()

    def reset(self) -> "LongestWinningStreak":
        """Clear observations, preserve input configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable observations processed by Rust."""
        return len(self._state)


__all__ = ["LongestWinningStreak"]

File: crates/taflow-metrics/src/metrics/longest_winning_streak.rs
use crate::{MetricError, MetricInputKind, MetricInputState, MetricResult, NanPolicy};

/// Longest consecutive run of strictly positive observations.
#[derive(Debug, Clone)]
pub struct LongestWinningStreak {
    input: MetricInputState,
    current_streak: usize,
    longest_streak: usize,
}

impl LongestWinningStreak {
    /// Construct an empty state for returns, raw period P&L, or closed trades.
    pub fn new(input_kind: MetricInputKind, nan_policy: NanPolicy) -> MetricResult<Self> {
        if !matches!(
            input_kind,
            MetricInputKind::Returns | MetricInputKind::RawPnl | MetricInputKind::Trades
        ) {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{input_kind:?}"),
                reason: "longest winning streak requires simple returns, raw period P&L, or closed-trade P&L",
            });
        }
        Ok(Self {
            input: MetricInputState::new(input_kind, nan_policy)?,
            current_streak: 0,
            longest_streak: 0,
        })
    }

    /// Append one observation and return the longest winning streak to date.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(observation) = self.input.append(value)? {
            if observation > 0.0 {
                self.current_streak += 1;
                self.longest_streak = self.longest_streak.max(self.current_streak);
            } else {
                self.current_streak = 0;
            }
        }
        Ok(self.value())
    }

    /// Append a chronological slice through the same persistent state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        for &value in values {
            self.append(value)?;
        }
        Ok(self.value())
    }

    /// Return the longest run length, or `None` when no observation exists.
    pub fn value(&self) -> Option<f64> {
        (!self.input.is_empty()).then_some(self.longest_streak as f64)
    }

    /// Return the current result without replaying prior observations.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving input configuration.
    pub fn reset(&mut self) {
        self.input.reset();
        self.current_streak = 0;
        self.longest_streak = 0;
    }

    /// Return the number of usable observations processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable observations have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}
